File: aiwaf/core/header_validation.py
import re

MAX_HEADER_BYTES = 32 * 1024
MAX_HEADER_COUNT = 100
MAX_USER_AGENT_LENGTH = 500
MAX_ACCEPT_LENGTH = 4096

REQUIRED_HEADERS = [
    'HTTP_USER_AGENT',
    'HTTP_ACCEPT',
]

BROWSER_HEADERS = [
    'HTTP_ACCEPT_LANGUAGE',
    'HTTP_ACCEPT_ENCODING',
    'HTTP_CONNECTION',
    'HTTP_CACHE_CONTROL',
]

SUSPICIOUS_USER_AGENTS = [
    r'bot', r'crawler', r'spider', r'scraper', r'curl', r'wget', r'python', r'java', r'node',
    r'go-http', r'axios', r'okhttp', r'libwww', r'lwp-trivial', r'mechanize', r'requests', r'urllib',
    r'httpie', r'postman', r'insomnia', r'^$', r'mozilla/4\.0$'
]

LEGITIMATE_BOTS = [
    r'googlebot', r'bingbot', r'slurp', r'duckduckbot', r'baiduspider', r'yandexbot',
    r'facebookexternalhit', r'twitterbot', r'linkedinbot', r'whatsapp', r'telegrambot',
    r'applebot', r'pingdom', r'uptimerobot', r'statuscake', r'site24x7'
]

SUSPICIOUS_COMBINATIONS = [
    {
        'condition': lambda headers: (
            headers.get('SERVER_PROTOCOL', '').startswith('HTTP/2') and
            'mozilla/4.0' in headers.get('HTTP_USER_AGENT', '').lower()
        ),
        'reason': 'HTTP/2 with old browser user agent'
    },
    {
        'condition': lambda headers: (
            headers.get('HTTP_USER_AGENT') and
            not headers.get('HTTP_ACCEPT')
        ),
        'reason': 'User-Agent present but no Accept header'
    },
    {
        'condition': lambda headers: (
            headers.get('HTTP_ACCEPT') == '*/*' and
            not any(h in headers for h in ['HTTP_ACCEPT_LANGUAGE', 'HTTP_ACCEPT_ENCODING'])
        ),
        'reason': 'Generic Accept header without language/encoding'
    },
    {
        'condition': lambda headers: (
            headers.get('HTTP_USER_AGENT') and
            not any(headers.get(h) for h in ['HTTP_ACCEPT_LANGUAGE', 'HTTP_ACCEPT_ENCODING', 'HTTP_CONNECTION'])
        ),
        'reason': 'Missing all browser-standard headers'
    },
    {
        'condition': lambda headers: (
            'HTTP_USER_AGENT' in headers and
            headers.get('SERVER_PROTOCOL') == 'HTTP/1.0' and
            'chrome' in headers.get('HTTP_USER_AGENT', '').lower()
        ),
        'reason': 'Modern browser with HTTP/1.0'
    }
]

def resolve_required_headers(config_required_headers, method=None):
    if config_required_headers is None:
        return list(REQUIRED_HEADERS)
    if isinstance(config_required_headers, (list, tuple)):
        return list(config_required_headers)
    if isinstance(config_required_headers, dict):
        if method:
            method_upper = method.upper()
            headers = config_required_headers.get(method_upper)
            if headers is not None:
                return list(headers)
        headers = config_required_headers.get("DEFAULT")
        if headers is not None:
            return list(headers)
    return list(REQUIRED_HEADERS)

def _check_user_agent(user_agent):
    if not user_agent:
        return None
        
    if len(user_agent) > MAX_USER_AGENT_LENGTH:
        return f"User-Agent longer than {MAX_USER_AGENT_LENGTH} chars"
    
    user_agent_lower = user_agent.lower()
    
    for legitimate_pattern in LEGITIMATE_BOTS:
        if re.search(legitimate_pattern, user_agent_lower):
            return None
            
    for suspicious_pattern in SUSPICIOUS_USER_AGENTS:
        if re.search(suspicious_pattern, user_agent_lower, re.IGNORECASE):
            return f"Pattern: {suspicious_pattern}"
            
    if len(user_agent) < 10:
        return "Too short"
        
    return None

def _check_header_combinations(headers, required_headers):
    if not required_headers:
        return None
    required = set(required_headers)
    for combo in SUSPICIOUS_COMBINATIONS:
        try:
            if combo.get('reason') == 'User-Agent present but no Accept header' and 'HTTP_ACCEPT' not in required:
                continue
            if combo['condition'](headers):
                return combo['reason']
        except Exception:
            continue
    return None

def _calculate_header_quality(headers):
    score = 0
    if headers.get('HTTP_USER_AGENT'):
        score += 2
    if headers.get('HTTP_ACCEPT'):
        score += 2
    for header in BROWSER_HEADERS:
        if headers.get(header):
            score += 1
    if headers.get('HTTP_ACCEPT_LANGUAGE') and headers.get('HTTP_ACCEPT_ENCODING'):
        score += 1
    if headers.get('HTTP_CONNECTION') == 'keep-alive':
        score += 1
    accept = headers.get('HTTP_ACCEPT', '')
    if 'text/html' in accept and 'application/xml' in accept:
        score += 1
    return score

def validate_headers_python(environ, method=None, config_required_headers=None, min_score=None):
    total_bytes = 0
    header_count = 0
    for key, value in environ.items():
        if not (key.startswith('HTTP_') or key in {'CONTENT_TYPE', 'CONTENT_LENGTH'}):
            continue

        header_count += 1
        value_str = value if isinstance(value, str) else str(value)
        total_bytes += len(key) + len(value_str)

        if total_bytes > MAX_HEADER_BYTES:
            return f"Header bytes exceed {MAX_HEADER_BYTES}"

    if header_count > MAX_HEADER_COUNT:
        return f"Header count exceeds {MAX_HEADER_COUNT}"

    user_agent = environ.get('HTTP_USER_AGENT', '')
    if user_agent and len(user_agent) > MAX_USER_AGENT_LENGTH:
        return f"User-Agent longer than {MAX_USER_AGENT_LENGTH} chars"

    accept_header = environ.get('HTTP_ACCEPT', '')
    if accept_header and len(accept_header) > MAX_ACCEPT_LENGTH:
        return f"Accept header longer than {MAX_ACCEPT_LENGTH} chars"

    required_headers = resolve_required_headers(config_required_headers, method)
    
    missing = []
    for header in required_headers:
        if not environ.get(header):
            missing.append(header.replace('HTTP_', '').replace('_', '-').lower())
    if missing:
        return f"Missing required headers: {', '.join(missing)}"

    suspicious_ua = _check_user_agent(user_agent)
    if suspicious_ua:
        return f"Suspicious user agent: {suspicious_ua}"

    suspicious_combo = _check_header_combinations(environ, required_headers)
    if suspicious_combo:
        return f"Suspicious headers: {suspicious_combo}"

    quality_score = _calculate_header_quality(environ)
    actual_min_score = min_score if min_score is not None else 3
    if required_headers and quality_score < actual_min_score:
        return f"Low header quality score: {quality_score}"

    return None

File: js/lib/headerValidation.js
let config = {
  enabled: false,
  requiredHeaders: [],
  blockedUserAgents: [],
  minScore: 3,
  maxHeaderBytes: 32 * 1024,
  maxHeaderCount: 100,
  maxUserAgentLength: 500,
  maxAcceptLength: 4096,
  suspiciousUserAgents: [],
  legitimateBots: [],
  staticExtensions: ['.css', '.js', '.png', '.jpg', '.jpeg', '.gif', '.ico', '.svg', '.woff', '.woff2', '.ttf'],
  staticPaths: ['/static/', '/media/', '/assets/', '/favicon.ico']
};

const BROWSER_HEADERS = [
  'accept-language',
  'accept-encoding',
  'connection',
  'cache-control'
];

const DEFAULT_SUSPICIOUS_UA = [
  /bot/i,
  /crawler/i,
  /spider/i,
  /scraper/i,
  /curl/i,
  /wget/i,
  /python/i,
  /java/i,
  /node/i,
  /go-http/i,
  /axios/i,
  /okhttp/i,
  /libwww/i,
  /lwp-trivial/i,
  /mechanize/i,
  /requests/i,
  /urllib/i,
  /httpie/i,
  /postman/i,
  /insomnia/i,
  /^$/,
  /^mozilla\/4\.0$/i
];

const DEFAULT_LEGIT_BOTS = [
  /googlebot/i,
  /bingbot/i,
  /slurp/i,
  /duckduckbot/i,
  /baiduspider/i,
  /yandexbot/i,
  /facebookexternalhit/i,
  /twitterbot/i,
  /linkedinbot/i,
  /whatsapp/i,
  /telegrambot/i,
  /applebot/i,
  /pingdom/i,
  /uptimerobot/i,
  /statuscake/i,
  /site24x7/i
];

function isStaticRequest(pathname) {
  const pathLower = String(pathname || '').toLowerCase();
  if (config.staticExtensions.some(ext => pathLower.endsWith(ext))) return true;
  if (config.staticPaths.some(prefix => pathLower.startsWith(prefix))) return true;
  return false;
}

function enforceHeaderCaps(headers) {
  let totalBytes = 0;
  let headerCount = 0;

  for (const [key, value] of Object.entries(headers)) {
    headerCount += 1;
    const valueStr = value === undefined || value === null ? '' : String(value);
    totalBytes += String(key).length + valueStr.length;

    if (totalBytes > config.maxHeaderBytes) {
      return `header_bytes_exceeded:${config.maxHeaderBytes}`;
    }
  }

  if (headerCount > config.maxHeaderCount) {
    return `header_count_exceeded:${config.maxHeaderCount}`;
  }

  const ua = headers['user-agent'];
  if (ua && String(ua).length > config.maxUserAgentLength) {
    return `user_agent_too_long:${config.maxUserAgentLength}`;
  }

  const accept = headers['accept'];
  if (accept && String(accept).length > config.maxAcceptLength) {
    return `accept_too_long:${config.maxAcceptLength}`;
  }

  return null;
}

function matchesPattern(list, value) {
  return list.some(pattern => pattern.test(value));
}

function checkSuspiciousCombinations(headers) {
  const protocol = String(headers['server-protocol'] || '').toUpperCase();
  const ua = String(headers['user-agent'] || '').toLowerCase();
  const accept = headers['accept'];

  if (protocol.startsWith('HTTP/2') && ua.includes('mozilla/4.0')) {
    return 'http2_with_old_ua';
  }

  if (headers['user-agent'] && !accept) {
    return 'ua_without_accept';
  }

  if (accept === '*/*' && !headers['accept-language'] && !headers['accept-encoding']) {
    return 'generic_accept_no_locale';
  }

  if (headers['user-agent'] && !headers['accept-language'] && !headers['accept-encoding'] && !headers['connection']) {
    return 'missing_browser_headers';
  }

  if (protocol === 'HTTP/1.0' && ua.includes('chrome')) {
    return 'modern_browser_http10';
  }

  return null;
}

function calculateHeaderQuality(headers) {
  let score = 0;
  const ua = headers['user-agent'];
  const accept = headers['accept'];

  if (ua) score += 2;
  if (accept) score += 2;

  for (const header of BROWSER_HEADERS) {
    if (headers[header]) score += 1;
  }

  if (headers['accept-language'] && headers['accept-encoding']) score += 1;

  if (String(headers['connection'] || '').toLowerCase() === 'keep-alive') score += 1;

  const acceptValue = String(accept || '').toLowerCase();
  if (acceptValue.includes('text/html') && acceptValue.includes('application/xml')) score += 1;

  return score;
}

module.exports = {
  init(opts = {}) {
    const requiredHeaders = opts.AIWAF_REQUIRED_HEADERS || [];
    const toRegexList = (items, fallback) => {
      if (!items) return fallback;
      if (!Array.isArray(items)) return fallback;
      return items.map(item => {
        if (item instanceof RegExp) return item;
        return new RegExp(String(item), 'i');
      });
    };
    config = {
      enabled: !!opts.AIWAF_HEADER_VALIDATION,
      requiredHeaders: Array.isArray(requiredHeaders)
        ? requiredHeaders.map(h => String(h).toLowerCase())
        : (requiredHeaders && typeof requiredHeaders === 'object' ? requiredHeaders : []),
      blockedUserAgents: (opts.AIWAF_BLOCKED_USER_AGENTS || []).map(u => String(u).toLowerCase()),
      minScore: Number.isFinite(Number(opts.AIWAF_HEADER_QUALITY_MIN_SCORE))
        ? Number(opts.AIWAF_HEADER_QUALITY_MIN_SCORE)
        : 3,
      maxHeaderBytes: Number.isFinite(Number(opts.AIWAF_MAX_HEADER_BYTES))
        ? Number(opts.AIWAF_MAX_HEADER_BYTES)
        : 32 * 1024,
      maxHeaderCount: Number.isFinite(Number(opts.AIWAF_MAX_HEADER_COUNT))
        ? Number(opts.AIWAF_MAX_HEADER_COUNT)
        : 100,
      maxUserAgentLength: Number.isFinite(Number(opts.AIWAF_MAX_USER_AGENT_LENGTH))
        ? Number(opts.AIWAF_MAX_USER_AGENT_LENGTH)
        : 500,
      maxAcceptLength: Number.isFinite(Number(opts.AIWAF_MAX_ACCEPT_LENGTH))
        ? Number(opts.AIWAF_MAX_ACCEPT_LENGTH)
        : 4096,
      suspiciousUserAgents: toRegexList(opts.AIWAF_SUSPICIOUS_USER_AGENTS, DEFAULT_SUSPICIOUS_UA),
      legitimateBots: toRegexList(opts.AIWAF_LEGITIMATE_BOTS, DEFAULT_LEGIT_BOTS),
      staticExtensions: opts.AIWAF_STATIC_EXTENSIONS || config.staticExtensions,
      staticPaths: opts.AIWAF_STATIC_PATHS || config.staticPaths
    };
  },

  validate(req) {
    if (!config.enabled) return null;

    const headers = Object.fromEntries(
      Object.entries(req.headers || {}).map(([key, value]) => [String(key).toLowerCase(), value])
    );
    headers['server-protocol'] = req.httpVersion ? `HTTP/${req.httpVersion}` : headers['server-protocol'];

    if (process.env.AIWAF_DEBUG_HEADERS) {
      console.error(`[HEADER-VALIDATION] headers keys: ${Object.keys(headers).join(', ')}`);
      console.error(`[HEADER-VALIDATION] user-agent: ${headers['user-agent']}`);
      console.error(`[HEADER-VALIDATION] accept: ${headers['accept']}`);
      console.error(`[HEADER-VALIDATION] accept-language: ${headers['accept-language']}`);
      console.error(`[HEADER-VALIDATION] accept-encoding: ${headers['accept-encoding']}`);
      console.error(`[HEADER-VALIDATION] connection: ${headers['connection']}`);
    }

    if (isStaticRequest(req.path || req.url || '')) return null;

    const capReason = enforceHeaderCaps(headers);
    if (capReason) {
      return capReason;
    }

    let requiredHeaders = config.requiredHeaders;
    if (requiredHeaders && typeof requiredHeaders === 'object' && !Array.isArray(requiredHeaders)) {
      const method = String(req.method || '').toUpperCase();
      requiredHeaders = requiredHeaders[method] || requiredHeaders.DEFAULT || [];
    }
    for (const header of requiredHeaders || []) {
      if (!headers[header]) {
        return `missing_header:${header}`;
      }
    }

    const ua = String(headers['user-agent'] || '').toLowerCase();
    if (config.legitimateBots.length && matchesPattern(config.legitimateBots, ua)) {
      return null;
    }
    if (config.suspiciousUserAgents.length && matchesPattern(config.suspiciousUserAgents, ua)) {
      return 'suspicious_user_agent';
    }
    if (config.blockedUserAgents.some(pattern => ua.includes(pattern))) {
      return 'blocked_user_agent';
    }

    const comboReason = checkSuspiciousCombinations(headers);
    if (comboReason) {
      return `suspicious_headers:${comboReason}`;
    }

    const minScore = (Array.isArray(requiredHeaders) && requiredHeaders.length === 0) ? 0 : config.minScore;
    if (minScore > 0) {
      const qualityScore = calculateHeaderQuality(headers);
      if (qualityScore < minScore) {
        return 'header_quality_low';
      }
    }

    return null;
  }
  ,
  getWasmConfig(method) {
    if (!config.enabled) return null;
    let requiredHeaders = config.requiredHeaders;
    if (requiredHeaders && typeof requiredHeaders === 'object' && !Array.isArray(requiredHeaders)) {
      const normalizedMethod = String(method || '').toUpperCase();
      requiredHeaders = requiredHeaders[normalizedMethod] || requiredHeaders.DEFAULT || [];
    }
    return {
      requiredHeaders: Array.isArray(requiredHeaders) ? requiredHeaders : [],
      minScore: (Array.isArray(requiredHeaders) && requiredHeaders.length === 0) ? 0 : config.minScore
    };
  },
  isStaticRequest(pathname) {
    return isStaticRequest(pathname);
  }
};
